src/xc_integrator/local_work_driver/device/scheme1_base.cxx

```diff
@@ -547,17 +667,71 @@ void AoSScheme1Base::eval_kern_exc_vxc_gga( const functional_type& func,
+                        1 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+        // Interleave gamma aa, bb, ab
+        stat = cudaMemcpy2D(base_stack.gamma_eval_device    , 3 * sizeof(double), base_stack.gamma_pp_eval_device,
+                        1 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+        stat = cudaMemcpy2D(base_stack.gamma_eval_device + 1, 3 * sizeof(double), base_stack.gamma_pm_eval_device,
+                        1 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+        stat = cudaMemcpy2D(base_stack.gamma_eval_device + 2, 3 * sizeof(double), base_stack.gamma_mm_eval_device,
+                        1 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+  }
   GauXC::eval_kern_exc_vxc_gga( func, data->total_npts_task_batch, 
     base_stack.den_eval_device, base_stack.gamma_eval_device, 
     base_stack.eps_eval_device, base_stack.vrho_eval_device, 
     base_stack.vgamma_eval_device, data->device_backend_->queue() );
 
-  hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1, 
-                    base_stack.weights_device, 1, base_stack.eps_eval_device, 1 );
-  hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1, 
+  if( is_RKS or is_excgrad ) {
+    hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1, 
                     base_stack.weights_device, 1, base_stack.vrho_eval_device, 1 );
-  hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1, 
+    hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1, 
                     base_stack.weights_device, 1, base_stack.vgamma_eval_device, 1 );
+  }
+  if( is_UKS ) {
+      // De-interleave pos/neg densities
+      auto stat        = cudaMemcpy2D(base_stack.vrho_pos_eval_device, 1 * sizeof(double), base_stack.vrho_eval_device,
+                        2 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+      stat             = cudaMemcpy2D(base_stack.vrho_neg_eval_device, 1 * sizeof(double), base_stack.vrho_eval_device + 1,
+                        2 * sizeof(double), 1 * sizeof(double), npts, cudaMemcpyDeviceToDevice);
+      // Multiply by weights point-by-point
+      hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1,
+                      base_stack.weights_device, 1, base_stack.vrho_pos_eval_device, 1 );
+      hadamard_product( data->device_backend_->master_blas_handle(), data->total_npts_task_batch, 1,
+                      base_stack.weights_device, 1, base_stack.vrho_neg_eval_device, 1 );
+
+      // De-interleave vgamma
+      stat            = cudaMemcpy2D(base_stack.vgamma_pp_eval_device, 1 * sizeof(double), base_stack.vgamma_eval_device,
```

Comment:
This can be done with one 2d copy if the arrays are stored contiguously  (they are)